src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1610 @@
+    // there is some other difference that prevents them from being 100%
+    // identical (for example, one might be an ExactType and the other a
+    // Literal; or both might be ExactTypes and only one might be nullable).
+    // In these cases we can emit a proper ExactType here, adding nullability
+    // if we need to.
+    value = ExactType(Type(
+      type.getHeapType(),
+      type.isNullable() || otherType.isNullable() ? Nullable : NonNullable));
+    return;
+  }
+
+  // Nothing else possible combines in an interesting way; emit a Many.
+  value = Many();
+}
+
+namespace {
+
+// We are going to do a very large flow operation, potentially, as we create
+// a Location for every interesting part in the entire wasm, and some of those
+// places will have lots of links (like a struct field may link out to every
+// single struct.get of that type), so we must make the data structures here
+// as efficient as possible. Towards that goal, we work with location
+// *indexes* where possible, which are small (32 bits) and do not require any
+// complex hashing when we use them in sets or maps.
+//
+// Note that we do not use indexes everywhere, since the initial analysis is
+// done in parallel, and we do not have a fixed indexing of locations yet. When
+// we merge the parallel data we create that indexing, and use indexes from then
+// on.
+using LocationIndex = uint32_t;
+
+#ifndef NDEBUG
+// Assert on not having duplicates in a vector.
+template<typename T> void disallowDuplicates(const T& targets) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  std::unordered_set<LocationIndex> uniqueTargets;
+  for (const auto& target : targets) {
+    uniqueTargets.insert(target);
+  }
+  assert(uniqueTargets.size() == targets.size());
+#endif
+}
+#endif
+
+// A link indicates a flow of content from one location to another. For
+// example, if we do a local.get and return that value from a function, then
+// we have a link from a LocalLocation to a ResultLocation.
+template<typename T> struct Link {
+  T from;
+  T to;
+
+  bool operator==(const Link<T>& other) const {
+    return from == other.from && to == other.to;
+  }
+};
+
+using LocationLink = Link<Location>;
+using IndexLink = Link<LocationIndex>;
+
+} // anonymous namespace
+
+} // namespace wasm
+
+namespace std {
+
+template<> struct hash<wasm::LocationLink> {
+  size_t operator()(const wasm::LocationLink& loc) const {
+    return std::hash<std::pair<wasm::Location, wasm::Location>>{}(
+      {loc.from, loc.to});
+  }
+};
+
+template<> struct hash<wasm::IndexLink> {
+  size_t operator()(const wasm::IndexLink& loc) const {
+    return std::hash<std::pair<wasm::LocationIndex, wasm::LocationIndex>>{}(
+      {loc.from, loc.to});
+  }
+};
+
+} // namespace std
+
+namespace wasm {
+
+namespace {
+
+// The data we gather from each function, as we process them in parallel. Later
+// this will be merged into a single big graph.
+struct CollectedInfo {
```

Comment:
If this only pertains to a single function, maybe name this `CollectedFuncInfo` or something?

Reply:
Sounds good, renamed.